src/nvim/os/job.c

```diff
@@ -117,63 +112,45 @@ Job *job_start(JobOptions opts, int *status)
   job->refcount = 1;
   job->stopped_time = 0;
   job->term_sent = false;
-  job->in = NULL;
-  job->out = NULL;
-  job->err = NULL;
   job->opts = opts;
   job->closed = false;
-
-  process_init(job);
-
-  if (opts.writable) {
-    handle_set_job((uv_handle_t *)job->proc_stdin, job);
-    job->refcount++;
-  }
-
-  if (opts.stdout_cb) {
-    handle_set_job((uv_handle_t *)job->proc_stdout, job);
-    job->refcount++;
-  }
-
-  if (opts.stderr_cb) {
-    handle_set_job((uv_handle_t *)job->proc_stderr, job);
-    job->refcount++;
-  }
+  job->in.closed = true;
+  job->out.closed = true;
+  job->err.closed = true;
 
   // Spawn the job
   if (!process_spawn(job)) {
-    if (opts.writable) {
-      uv_close((uv_handle_t *)job->proc_stdin, close_cb);
+    if (job->opts.writable) {
+      uv_close((uv_handle_t *)job->proc_stdin, NULL);
     }
-    if (opts.stdout_cb) {
-      uv_close((uv_handle_t *)job->proc_stdout, close_cb);
+    if (job->opts.stdout_cb) {
+      uv_close((uv_handle_t *)job->proc_stdout, NULL);
     }
-    if (opts.stderr_cb) {
-      uv_close((uv_handle_t *)job->proc_stderr, close_cb);
+    if (job->opts.stderr_cb) {
+      uv_close((uv_handle_t *)job->proc_stderr, NULL);
     }
     process_close(job);
     event_poll(0);
```

Comment:
`event_poll(0)` should be called before `process_close(job)`, to prevent freeing the job structure too soon.
Without this change the following patch (simulates an error while spawning a pty) and usage of `:term` should produce an crash:
```diff
diff --git a/src/nvim/os/pty_process.c b/src/nvim/os/pty_process.c
index cb59355..ebffac1 100644
--- a/src/nvim/os/pty_process.c
+++ b/src/nvim/os/pty_process.c
@@ -72,6 +72,7 @@ bool pty_process_spawn(Job *job) FUNC_ATTR_NONNULL_ALL
     init_child(job);
     abort();
   }
+  goto error;
 
   // make sure the master file descriptor is non blocking
   int master_status_flags = fcntl(master, F_GETFL);
```